### ISSUE TITLE
Incompatibility between embed proxying and X-Frames: SAMEORIGIN

### ISSUE THREAD

**Issue by MrPetovan** [friendica/friendica#4155]
I'm having a display issue where embeds frames from Wordpress won't show in my feed because Wordpress itself uses a frame whose headers comprises `X-Frames: SAMEORIGIN`. This means the browser won't load the frame if the parent frame isn't the same exact domain. Normally, it should always work, as the parent embed URL is supposed to be on the same Wordpress domain.

Except with Friendica proxy feature. Since we proxy the embed URL through the Friendica domain, the browser enforces the policy and don't display the frame.

Example:

Original post with WP embed: https://friendica.mrpetovan.com/display/735a2029185a48fca2c460a319316270
Working embed URL: https://blog.mrpetovan.com/slice-of-life/2017-holiday-road-trip-florida-new-orleans/embed/
Non-working proxy URL: https://friendica.mrpetovan.com/oembed/aHR0cHM6Ly9ibG9nLm1ycGV0b3Zhbi5jb20vc2xpY2Utb2YtbGlmZS8yMDE3LWhvbGlkYXktcm9hZC10cmlwLWZsb3JpZGEtbmV3LW9ybGVhbnMv

Wordpress embed URL code:

````html
<html>
  <body>
    <blockquote class="wp-embedded-content">
      <a href="https://blog.mrpetovan.com/slice-of-life/2017-holiday-road-trip-florida-new-orleans/">A (Mostly) Sunny Holiday Road-Trip Through Florida To New Orleans</a>
    </blockquote>
    <script type='text/javascript'>
    <!--//--><![CDATA[//><!--
        ...
    //--><!]]>
    </script>
    <iframe
      sandbox="allow-scripts"
      security="restricted"
      src="https://blog.mrpetovan.com/slice-of-life/2017-holiday-road-trip-florida-new-orleans/embed/#?secret=..."
      width="425"
      height="240"
      title="&#8220;A (Mostly) Sunny Holiday Road-Trip Through Florida To New Orleans&#8221; &#8212; MrPetovan.com"
      frameborder="0"
      marginwidth="0"
      marginheight="0"
      scrolling="no"
      class="wp-embedded-content">
    </iframe>
  </body>
</html>
````

The solution is simple on paper: stop proxying WordPress embeds. However, detecting WordPress wouldn't be trivial, or it would have to be manually by domain, which would be cumbersome. The last resort could be to disable proxy for all embeds, but then the browser's Mixed Content policy would kick in and refuse to display non-https embeds.

Or maybe just proxying non-http embeds that can be identified just on the URL alone?

What do you think?

**Comment by annando** [comment]
Or we disable rich oembed contents?

**Comment by rabuzarus** [comment]
I would rather vote for disabling rich oembed content and implementing a whitelist to enable rich oembed contents for some specific domains. This whitelist should be expandable by the admin.

**Comment by MrPetovan** [comment]
I don't understand your answers, the initial issue never was about enabling/disabling rich oembed content, only the proxying part of it. I want to keep rich OEmbed content because it looks nice, and I want it to look nice no matter the remote domain.

There was a known mixed content issue that was solved by the proxy, but proxying HTTPS embeds as well introduced a new issue with WordPress embeds. The solution is to disable proxying (and not all rich OEmbed content) for HTTPS links. The issue would then be cornered to non-HTTPS Wordpress embeds but then the cost to fix them becomes prohibitive.

**Comment by MrPetovan** [comment]
It isn't a security concern, it isn't a performance concern, it just is a display quirk introduced by one of our own features.

**Comment by annando** [comment]
The proxy is also a way to disable tracking attempts.

**Comment by rabuzarus** [comment]
@MrPetovan 
Sry. My post was just an answer to @annando comment and not about your issue.
I also like some rich oembed content. That's why I'm against disabling it completely (but in many cases rich oembed content could be annoying). But I stop here because it is not about the issue ;-)

**Comment by MrPetovan** [comment]
@annando I get your point, but I don't think it should harm normal display, and in this case it does.

**Comment by MrPetovan** [comment]
While I was reading https://oembed.com/#section3, I came upon this:
> When a consumer displays HTML (as with video embeds), there's a vector for XSS attacks from the provider. To avoid this, it is recommended that consumers display the HTML in an iframe, hosted from another domain. This ensures that the HTML cannot access cookies from the consumer domain.

When we proxy OEmbed from the same Friendica domain, we actually are leaving the Friendica cookie accessible to third-party websites.

I strongly recommend we drop the proxying altogether for OEmbed content.

**Comment by annando** [comment]
Okay. But I would like to include the points that where mentioned here. Means: When proxying is bad, we can drop it - but when reworking that part, we could include an optional whitelist functionality that could be combined with the already existing switch that disables rich content at all.

Example:

```
[x] disable rich content in OEmbed except for whitelisted domains.
Whitelist: ...
```

**Comment by MrPetovan** [comment]
I see what you mean, and it's possible, although I will probably use a newline-delimited list, it's way more readable in a textarea than a comma-separated list.

**Comment by annando** [comment]
I totally rely on your decision how to implement it.

**Comment by MrPetovan** [comment]
The other place I can think of is the blockem block list setting, and it was cumbersome to edit manually, I even added a brand-new admin page just to circumvent that! Comma-separated lists should be left for computer-to-computer communication.

**Comment by annando** [comment]
There are more places. For example you can enter multiple mail addresses for the admin. And the list of keywords that should be ignored when doing "remote self", is comma separated as well.

I guess I'm hearing another task waiting for you :-)

**Comment by MrPetovan** [comment]
*sigh*

I guess so.

**Comment by MrPetovan** [comment]
I'm confused about the whitelist, should it be about OEmbed at all, or just the proxying part?

**Comment by annando** [comment]
I would vote for an OEmbed whitelist.

**Comment by MrPetovan** [comment]
Ok good, it makes more sense to me.
